src/ServiceControl.Monitoring/Http/Diagrams/MonitoredEndpointMessageTypeParser.cs

```diff
@@ -12,13 +12,15 @@ public class MonitoredEndpointMessageTypeParser
 
         public MonitoredEndpointMessageType Parse(string typeName)
         {
-            if (!messageTypes.ContainsKey(typeName))
-            {
-                var messageType = FromTypeName(typeName);
+            MonitoredEndpointMessageType messageType;
 
-                messageTypes.TryAdd(typeName, messageType);
+            if (messageTypes.TryGetValue(typeName, out messageType))
+            {
+                return messageType;
             }
 
+            messageTypes.TryAdd(typeName, FromTypeName(typeName));
+
             return messageTypes[typeName];
```

Comment:
If that's a cached item, you could return also the one that you `TryAdd` and remove accessing dictionary altogether. This is not heavy requirement as the hot path (reading) issue has been addressed.